pkg/gitserver/controller/controller.go

```diff
@@ -0,0 +1,231 @@
+		// remove the reference from the git repository
+		if err := c.removeFromRepository(entry.Repository, bcEntry(bc.Name, entry.Reference)); err != nil {
+			return err
+		}
+		delete(c.BCCache, bc.Name)
+	}
+
+	if !inCache || removed {
+		// If not already in the cache or it has been removed because of a mismatch, add to the
+		// repository config
+		if err := c.addToRepository(repoName, bcEntry(bc.Name, bc.Spec.Source.Git.Ref)); err != nil {
+			return err
+		}
+		c.BCCache[bc.Name] = BCCacheEntry{
+			Repository: repoName,
+			Reference:  bc.Spec.Source.Git.Ref,
+		}
+	}
+
+	return nil
+}
+
+func (c *BuildConfigController) HandleBuildConfigDelete(bc *buildapi.BuildConfig) error {
+	// First remove from the cache (and corresponding repository) if present
+	entry, inCache := c.BCCache[bc.Name]
+	if inCache {
+		if err := c.removeFromRepository(entry.Repository, bcEntry(bc.Name, entry.Reference)); err != nil {
+			return err
+		}
+		delete(c.BCCache, bc.Name)
+	}
+	// Ensure that even if it's not in the cache, that its
+	// corresponding repository is updated if there is one.
+	repoName, hasGitServerSource := c.gitServerRepository(bc)
+	if hasGitServerSource {
+		if !inCache || entry.Repository != repoName || entry.Reference != bc.Spec.Source.Git.Ref {
+			if err := c.removeFromRepository(repoName, bcEntry(bc.Name, bc.Spec.Source.Git.Ref)); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *BuildConfigController) gitServerRepository(bc *buildapi.BuildConfig) (name string, isRepo bool) {
+	if bc.Spec.Source.Git == nil {
+		return "", false
+	}
+	parsedURL, err := url.Parse(bc.Spec.Source.Git.URI)
+	if err != nil {
+		return "", false
+	}
+	if parsedURL.Host != c.PublicURL.Host ||
+		parsedURL.Scheme != c.PublicURL.Scheme {
+		return "", false
+	}
+	path := parsedURL.Path
+	if c.PublicURL.Path != "" {
+		if !strings.HasPrefix(path, c.PublicURL.Path) {
+			return "", false
+		}
+		path = strings.TrimPrefix(path, c.PublicURL.Path)
+	}
+	path = strings.Trim(path, "/")
+	parts := strings.Split(path, "/")
+	if len(parts) > 1 {
+		return "", false
+	}
+	if !strings.HasSuffix(path, ".git") {
+		path += ".git"
+	}
+	return path, true
+}
+
+func (c *BuildConfigController) repoDir(name string) (string, error) {
+	dir := filepath.Join(c.Home, name)
+	info, err := os.Stat(dir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return "", nil
+		}
+		return "", err
+	}
+	if !info.IsDir() {
+		return "", fmt.Errorf("%s is not a directory", dir)
+	}
+	return dir, nil
+}
+
+func (c *BuildConfigController) addToRepository(repo, bcref string) error {
+	glog.Infof("Adding BuildConfig reference %s to repository %s", bcref, repo)
+	dir, err := c.repoDir(repo)
+	if err != nil {
+		return err
+	}
+	if len(dir) == 0 {
+		dir, err = c.createRepo(repo)
+		if err != nil {
+			return err
+		}
+	}
+	cmd := exec.Command("git", "config", "--add", "openshift.io.buildconfig", bcref)
+	cmd.Dir = dir
+	return cmd.Run()
```

Comment:
(i prefer to have some helper/utility function to execute git command, capture output and in case of error, translate that output into a error message)